### PR TITLE
Implement project configurator

### PR DESCRIPTION
**Pull request by jankeromnes**
Initial pass to fix https://github.com/gitpod-io/gitpod/issues/4648

Depends on & contains https://github.com/gitpod-io/gitpod/pull/4846

What it does:

- Adds a `config` (text) field to `d_b_project`, so that Projects can have a custom `.gitpod.yml` content
- Adds a Project Configurator view in the dashboard, as final step in the Add Project flow (also currently accessible via a Project's UI)
- When determining the config of a new workspace, use this priority order:
    1. From repository
    2. (_new_) From Project DB
    3. From definitely-gp
    4. Derived (Go) or default
- When starting from a custom `.gitpod.yml` (from the Project DB), the initializer also places the uncomitted `.gitpod.yml` into the workspace (to encourage repo-based configs)

How to test:

- Create a new team
- Add a new project (either by inserting into `d_b_project`, or we set up a GitHub App integration -- @AlexTugarev could you help with the latter?)
- If the project already has a repo-based config, the configurator should indicate that and be disabled. If there is no repo config, the configurator should allow you to write a new `.gitpod.yml` (e.g. based on one of the presets) and then test it / iterate on it in a quick loop (with the build output on the right)

The UX of this is still rough, but the goal is to add this final step rather soon, and then iterate on the entire Add Project flow.

### CODE REVIEW

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (jankeromnes) [inline components/server/ee/src/workspace/workspace-factory.ts:53]: FIXME: Move into configProvider + respect priority order (from repo > from project).

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4631#" title="Author self-approved">jankeromnes</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4631#" title="Author self-approved">jankeromnes</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jankeromnes** [comment]
🎰

/werft run

**Comment by AlexTugarev** [comment]
Happy to help with the setup of the GitHub App!

0. Create one on GitHub for your branch. Store the private key. 
1. Start a workspace in that `kubectx`.
2. Modify `value.yaml` and add...
```
    github:
      app:
        enabled: true
        appId: **********
        webhookSecret: ***************
        appCert: server-github-app-cert
        marketPlaceName: gitpod-io
```
3. Redeploy with that change: `werft run github -s chart/values.yaml -f`
4. Upload PK: 
```
kubectl create secret generic server-github-app-cert --from-literal=cert="-----BEGIN RSA PRIVATE KEY-----
...
```
Which might require to delete the server pod depending on the timing of the re-deployment.

File: components/dashboard/src/components/MonacoEditor.tsx
Comment (AlexTugarev) [inline components/dashboard/src/components/MonacoEditor.tsx:15]: `if (!containerRef.current)` ?

Comment (jankeromnes) [reply components/dashboard/src/components/MonacoEditor.tsx:15]: Hmm, I believe this is correct as is: You want the containing element (`containerRef.current`) to exist, so that you can mount a new editor (`editorRef.current`) on it.

Comment (AlexTugarev) [reply components/dashboard/src/components/MonacoEditor.tsx:15]: oops! I misread the refs and confused editorRef with it.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1495]: Why not `checkAndBlockUser`? IMO we block all modifying/creating actions on block.

**Comment by jankeromnes** [comment]
@AlexTugarev @geropl @svenefftinge Please approve or reject this Pull Request when you have time. 🙂 (Or, please let me know if you have any questions! Happy to jump on a call to walk through this.)

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (AlexTugarev) [inline components/dashboard/src/components/PrebuildLogs.tsx:124]: we've just merged some related changes on main.
actually, you'd be a better reviewer for it. 🙏🏻

Comment (jankeromnes) [reply components/dashboard/src/components/PrebuildLogs.tsx:124]: Ah, true, just saw the conflict.

I'm simply moving this method into a different file, because unlazily importing from `WorkspaceLogs.tsx` pulls in Xterm.js into all dashboard pages. 🙂

Will fix the conflict.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: Agreed. This initial cut-through doesn't handle Dockerfile, but I was planning to look into potentially storing a custom Dockerfile as well in the near future.

(For now, I think it adds too much complexity to a simple configurator -- ideally I'd even want to remove the YAML editor entirely, and just have people click together a simple config without "scaring them away".)

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:34]: Out of curiosity, when do we want to delete a project? Is this a feature we're planning to support in Gitpod's dashboard?

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: I generally agree with keeping it simple, but in this case we should think a bit ahead as we are talking about DB migrations.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1500]: Note: I've just moved this function.

@AlexTugarev do you still need this, or can I remove the dead code for now?

File: components/server/src/workspace/workspace-factory.ts
Comment (jankeromnes) [reply components/server/src/workspace/workspace-factory.ts:141]: Good point, thanks. However, is that still actually a problem when the Project DB has an index on cloneURL?

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: > ideally I'd even want to remove the YAML editor entirely, and just have people click together a simple config without "scaring them away

The config form should really just be a view on top of the yaml.

File: components/server/src/workspace/workspace-factory.ts
Comment (svenefftinge) [reply components/server/src/workspace/workspace-factory.ts:141]: I think so, yes. I cannot tell without measuring how much it contributes to the startup performance, though. But this really is the most important path, that is already way too long. We can keep it like it is now and do an optimization pass later.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:34]: If I have deleted the repo, for instance.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (geropl) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: What about a `JSON` with key being the path, value the content? Basically mimicking a file system overlay.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (geropl) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:34]: noted: we used `markDeleted` or `softDeleted` for thus "user-deleted" entities.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: I think this should be 
```
@Column("simple-json") 
config?: {[filePath: string]: string}
```
And on the db maybe even `mediumtext`

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: Alternatively, since we only need a .gitpod.yml and a Dockerfile, we could also have something like `config?: { configString: string, dockerfile?: ImageConfigFile }`

See the `ImageConfigFile` type here:

https://github.com/gitpod-io/gitpod/blob/313fc79b91ac9fe9d17054beac363c8db43555d4/components/gitpod-protocol/src/protocol.ts#L746-L751

This would avoid all the ambiguities around `gitpod.yml` vs `.gitpod.yml` vs `gitpod.yaml` vs `.gitpod/gitpod.yaml` etc, and I think there is no need to implement a "generic file system as JSON object" here.

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: We should solve that on the UI side and be flexible on the DB level. Going forward we might allow alternatives or just change the pattern.

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: It also aligns with the AdditionalContent shape

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:1500]: as long as github-app.ts is in EE space, we need to fill in this gap here. 
there should have been a comment telling "implemented in EE"

Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1500]: Ah, this is empty in non-EE code. I'm still not sure we want to keep this as is though.

Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:1500]: Last time we discussed this, there was no decision made about moving the code. What do you suggest?

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: I've implemented this change, but I find the data model less elegant.

The config string is the primary thing we care about in many places, but now to get it, we have to use the "magical" `'.gitpod.yml'` key everywhere in combination with the `project.config` object, with possible typos and increased opportunities to get `undefined`.

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:30]: Fair point. We can make the type not a string but a literal type for now and either add additional values when needed or soften it to string later.

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts]: something like this:
```
type ConfigPath = '.gitpod.yml' | '.gitpod.Dockerfile';
@Column("simple-json", { nullable: true }) 
config?: {[filePath: ConfigPath]: string}
```

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:1517]: Strangely, the dashboard now fails to load projects, and shows this in the console:

```
Uncaught (in promise) Error: Request getProjects failed with message: this.projectDB.trace is not a function
```

Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1517]: I'm calling this a bad deployment.

File: components/gitpod-db/src/typeorm/entity/db-project.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project.ts]: Hmm, TypeScript didn't seem to like the union type for `filePath`.

I think I'll go with this instead for now (easy to extend later):

```
  @Column("simple-json", { nullable: true })
  config?: { '.gitpod.yml': string };
```

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1517]: Re-deploying didn't help 😕

Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts]: Removing the tracing for now: It doesn't work.

Also, it doesn't really make sense to trace an extremely small, new DB, while all the other DBs are not traced (except WorkspaceDB, the biggest one, where tracing makes the most sense).

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (jankeromnes) [inline components/dashboard/src/components/PrebuildLogs.tsx:45]: @geropl Please double-check here: Is this really what the `checkIfDone` function is supposed to do?

(I'm not exactly sure yet how `watchHeadlessLogs` works exactly.)

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (jankeromnes) [inline components/dashboard/src/components/PrebuildLogs.tsx:124]: Note: I'm moving this function out of `WorkspaceLogs.tsx` and into the new component `PrebuildLogs.tsx`, because it is imported eagerly in various places.

However, we don't want to import `WorkspaceLogs.tsx` eagerly, because that file imports all of Xterm.js (you'll notice that `WorkspaceLogs` itself is always imported lazily -- same for `MonacoEditor.tsx` now).

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (geropl) [reply components/dashboard/src/components/PrebuildLogs.tsx:45]: @jankeromnes Yes, if `workspaceInstance` is updated properly (I assume it is).

Ideally we'd have a API for "getPrebuildState" so we don't have to interpret this from the underlying instance, but the result should be the same.

Comment (jankeromnes) [reply components/dashboard/src/components/PrebuildLogs.tsx:45]: Thanks! (In the future, I think it would be cool if `watchHeadlessLogs` can figure out on its own whenever the job it watches is done or not 😇)

File: components/server/src/workspace/workspace-factory.ts
Comment (jankeromnes) [reply components/server/src/workspace/workspace-factory.ts:141]: So, I think we either fetch the project twice (as currently), or we include the entire `ProjectConfig` in the `WorkspaceConfig` shape (and thus store a config multiple times in DB, for every new workspace).

Not sure what the lesser of evils is, or if maybe there is a better way.

Comment (svenefftinge) [reply components/server/src/workspace/workspace-factory.ts:141]: We can keep it like it is now and do an optimization pass later.

Comment (jankeromnes) [reply components/server/src/workspace/workspace-factory.ts:141]: I almost had a single project fetch in `configProvider`, with the project config stored in `WorkspaceConfig`, however the code below also needs the project ID. 🤦

So maybe it could be done the other way around, passing the Project to the configProvider, but I 1000% agree about doing optimization passes on workspace start later (soon).

Comment (jankeromnes) [reply components/server/src/workspace/workspace-factory.ts:144]: I've added a `TODO` comment at this location.

**Comment by jankeromnes** [comment]
Alright, I think this could be merged as is.

@svenefftinge and @AlexTugarev could you please take another look?

### How to test

1. Join the "Gitpod" Team using [this link](https://jx-configure-project.staging.gitpod-dev.com/teams/join?inviteId=23b7939c-b287-4e0e-9f63-925f3d085e3e)
2. In the [team projects](https://jx-configure-project.staging.gitpod-dev.com/gitpod/projects), open any of the 2 projects, then open the "Configure" tab:
    - For the "gitpod" project, a config is detected in the repo, so the editor should be read-only
    - For the "test" project (empty repository), you should be able to pick a configuration preset and then test it

Note: There is no GitHub App/Integration for this test environment. The Project Configurator is setup as the last step in the wizard, but I couldn't test the entire flow yet (I only tested the configurator specifically).

### Follow-ups

A few review discussions are still open. I think it's best to address them in follow-up issues & Pull Requests, in order to get this first pass merged sooner and spend more time iterating on the "wizard" flow holistically.

Still, please do feel free to share thoughts & feedback, and anything bigger than a quick fix before merging will become a follow-up issue.

**Review by svenefftinge — COMMENTED** [review]
There seems to be an issue with the prebuild logs. They logs stay empty for me and I could only see them after reloading. 
Try with: https://jx-configure-project.staging.gitpod-dev.com/#prebuild/https://github.com/gitpod-io/spring-petclinic
Could it be because of the changes in the PR?

**Comment by jankeromnes** [comment]
> There seems to be an issue with the prebuild logs. They logs stay empty for me and I could only see them after reloading.
> Try with: https://jx-configure-project.staging.gitpod-dev.com/#prebuild/https://github.com/gitpod-io/spring-petclinic
> Could it be because of the changes in the PR?

Hmm, I also no longer see prebuild logs, in the "manual prefix" page or in the "Prebuild in Progress" page.

Also, I got this error twice in the console:

```
re-trying headless-logs because: error while listening to stream Error: Headless logs for f1ceab3a-3678-4607-ab74-9f070b39f406 not found
```

Then the prebuild was eventually successfully finished, and started in a new workspace, although I didn't get any logs.

I don't know if this can be related to this PR, or if this is a problem with dev-staging (I'm under the impression that headless logs often simply don't work in dev-staging, while they do work better in staging & production).

Are other dev-staging deployments affected?

**Review by svenefftinge — APPROVED** [review]

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 2f6484b357b0c46d8754a426ba813d1eb6df81c5</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4631#" title="Author self-approved">jankeromnes</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4631#pullrequestreview-712547842" title="Approved">svenefftinge</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes,svenefftinge]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes,svenefftinge]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes,svenefftinge]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes,svenefftinge]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes,svenefftinge]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
@gtsiolis: [dog image](https://random.dog/61315d88-8e08-46ff-996b-5d6c2696c14d.gif)

<details>

In response to [this](https://github.com/gitpod-io/gitpod/pull/4631#issuecomment-884824936):

>/woof


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Pull request merged**
